File: Capítulo_13/Algoritmos de ordenação/Radix sort/Código em Python/radixSort.py
# Algoritmo de radix sort:
def radix_sort(array, radix_base = 10):
    if len(array) < 2:
        return array
    
    min_value = find_min_value(array)
    max_value = find_max_value(array)
    significant_digit = 1

    while (max_value - min_value) / significant_digit >= 1:
        array = counting_sort_for_radix(array, radix_base, significant_digit, min_value)
        significant_digit *= radix_base
    
    return array

# Função find_min_value
def find_min_value(array):
    min = array[0]
    for i in range(len(array)):
        if array[i] < min:
            min = array[i]

    return min

# Função find_max_value
def find_max_value(array):
    max = array[0]
    for i in range(len(array)):
        if array[i] > max:
            max = array[i]
        
    return max

# Função counting_sort_for_radix
def counting_sort_for_radix(array, radix_base, significant_digit, min_value):
    buckets_index = None
    buckets = [0 for _ in range(radix_base)]
    aux = [0] * len(array)

    for i in range(len(array)):
        buckets_index = (array[i] - min_value) // significant_digit % radix_base
        buckets[buckets_index] += 1
    
    for i in range(1, radix_base):
        buckets[i] += buckets[i - 1]

    for i in range(len(array) - 1, -1, -1):
        buckets_index = (array[i] - min_value) // significant_digit % radix_base
        aux[buckets[buckets_index] - 1] = array[i]
        buckets[buckets_index] -= 1
        
    for i in range(len(array)):
        array[i] = aux[i]

    return array

# Teste da função radixSort
array = [ 456, 789, 123, 0, 1, 32, 4, 243, 321, 42, 90, 10, 999]
print(", ".join(map(str, array))) # 456, 789, 123, 0, 1, 32, 4, 243, 321, 42, 90, 10, 999
array = radix_sort(array)
print(", ".join(map(str, array))) # 0, 1, 4, 10, 32, 42, 90, 123, 243, 321, 456, 789, 999

File: Capítulo_13/Algoritmos de ordenação/Radix sort/radixSort.js
// Algoritmo de radix sort:
function radixSort(array, radixBase = 10) {
    if (array.length < 2) {
        return array;
    }
    const minValue = findMinValue(array);
    const maxValue = findMaxValue(array);
    let significantDigit = 1;
    while ((maxValue - minValue) / significantDigit >= 1) {
        array = countingSortForRadix(
            array,
            radixBase,
            significantDigit,
            minValue
        );
        significantDigit *= radixBase;
    }
    return array;
}

// Função findMinValue
function findMinValue(array) {
    let min = array[0];
    for (let i = 0; i < array.length; i++) {
        if (array[i] < min) {
            min = array[i];
        }
    }
    return min;
}

// Função findMaxValue
function findMaxValue(array) {
    let max = array[0];
    for (let i = 0; i < array.length; i++) {
        if (array[i] > max) {
            max = array[i];
        }
    }
    return max;
}

// Função countingSortForRadix
function countingSortForRadix(array, radixBase, significantDigit, minValue) {
    let bucketsIndex;
    const buckets = [];
    const aux = [];
    for (let i = 0; i < radixBase; i++) {
        buckets[i] = 0;
    }
    for (let i = 0; i < array.length; i++) {
        bucketsIndex = Math.floor(
            ((array[i] - minValue) / significantDigit) % radixBase
        );
        buckets[bucketsIndex]++;
    }
    for (let i = 1; i < radixBase; i++) {
        buckets[i] += buckets[i - 1];
    }
    for (let i = array.length - 1; i >= 0; i--) {
        bucketsIndex = Math.floor(
            ((array[i] - minValue) / significantDigit) % radixBase
        );
        aux[--buckets[bucketsIndex]] = array[i];
    }
    for (let i = 0; i < array.length; i++) {
        array[i] = aux[i];
    }
    return array;
}

// Teste da função radixSort
let array = [ 456, 789, 123, 0, 1, 32, 4, 243, 321, 42, 90, 10, 999];
console.log(array.join(", ")); // 456, 789, 123, 0, 1, 32, 4, 243, 321, 42, 90, 10, 999
array = radixSort(array);
console.log(array.join(", ")); // 0, 1, 4, 10, 32, 42, 90, 123, 243, 321, 456, 789, 999
